### DISCUSSION TITLE
Proper way to recover if scheduler process crashes?

### DISCUSSION THREAD

**Discussion by tparvi** [quartznet/quartznet#2049] [Q&A]
We are running quartz.net in kubernetes which means that sometimes the scheduler process needs to be killed while jobs are running. We are also storing job configurations inside appsettings.json and when scheduler starts it will reschedule existing jobs (schedule might have changed) and schedule new jobs that are not yet know to quartz.

When the process crashes it seems the `qrtz_triggers` table will contain row where the name of the trigger is like `MT_8939...` and group is `DEFAULT` as seen in the following:

![image](https://github.com/quartznet/quartznet/assets/871053/52ec9ccb-02c8-4591-80bc-b911ce3cfde8)

Since we are rescheduling jobs at startup, the current solution is to reschedule only triggers that belong to our group:

```
private static async Task RescheduleExistingJob(IScheduler scheduler, JobConfiguration config,
	JobKey registeredJob)
{
	var jobTriggers = await scheduler.GetTriggersOfJob(registeredJob);
	if (!jobTriggers.Any())
	{
		return;
	}
	
	foreach (var trigger in jobTriggers)
	{
		// Do not reschedule jobs which do not belong to our group.
		// They are most likely failed jobs which quarz will clean up
		if (trigger.Key.Group != JobConfiguration.GroupName)
		{
			continue;
		}
			
		var triggerBuilder = trigger.GetTriggerBuilder();
		triggerBuilder.WithCronSchedule(config.Schedule);
		await scheduler.RescheduleJob(trigger.Key, triggerBuilder.Build());
	}
}

```

This seem to work and quartz will clean up the completed triggers:

```
ClusterManager: detected 1 failed or restarted instances.
ClusterManager: Scanning for instance "PtvScheduler-d6dc42f6-58d9-47e4-b81e-2f45a27c7707"'s failed in-progress jobs.
ClusterManager: ......Deleted 1 complete triggers(s).
ClusterManager: ......Cleaned-up 1 other failed job(s).
Scheduler Ptv Scheduler_$_PtvScheduler-894533a7-2713-41f8-813a-e3c2cc5b20b4 started.
```

Is this correct way to handle the situation or should be do something else?

**Comment by lahma** [comment]
One thing that you could consider is using the [Quartz's native XML configuration](https://github.com/quartznet/quartznet/tree/main/src/Quartz.Examples/15_ConfigureJobSchedulingByUsingXmlConfigurations). It handles checking of new and missing items. You can check how [XMLSchedulingDataProcessor](https://github.com/quartznet/quartznet/blob/main/src/Quartz/Xml/XMLSchedulingDataProcessor.cs) works for some ideas how to sync.

Generally if you have a fixed set of triggers and jobs you should try to name them with static ids (group + name) instead of Quartz generated dynamic ones so that it's easier to discover desired state.
